src/test/blockfilter_index_tests.cpp

```diff
@@ -0,0 +1,307 @@
+{
+    const CChainParams& chainparams = Params();
+    std::unique_ptr<CBlockTemplate> pblocktemplate = BlockAssembler(chainparams).CreateNewBlock(scriptPubKey);
+    CBlock& block = pblocktemplate->block;
+    block.hashPrevBlock = prev->GetBlockHash();
+    block.nTime = prev->nTime + 1;
+
+    // Replace mempool-selected txns with just coinbase plus passed-in txns:
+    block.vtx.resize(1);
+    for (const CMutableTransaction& tx : txns) {
+        block.vtx.push_back(MakeTransactionRef(tx));
+    }
+    // IncrementExtraNonce creates a valid coinbase and merkleRoot
+    unsigned int extraNonce = 0;
+    IncrementExtraNonce(&block, prev, extraNonce);
+
+    while (!CheckProofOfWork(block.GetHash(), block.nBits, chainparams.GetConsensus())) ++block.nNonce;
+
+    return block;
+}
+
+static bool BuildChain(const CBlockIndex* pindex, const CScript& coinbase_script_pub_key,
+                       size_t length, std::vector<std::shared_ptr<CBlock>>& chain)
+{
+    std::vector<CMutableTransaction> no_txns;
+
+    chain.resize(length);
+    for (auto& block : chain) {
+        block = std::make_shared<CBlock>(CreateBlock(pindex, no_txns, coinbase_script_pub_key));
+        CBlockHeader header = block->GetBlockHeader();
+
+        CValidationState state;
+        if (!ProcessNewBlockHeaders({header}, state, Params(), &pindex, nullptr)) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+BOOST_FIXTURE_TEST_CASE(blockfilter_index_initial_sync, TestChain100Setup)
+{
+    BlockFilterIndex filter_index(BlockFilterType::BASIC, 1 << 20, true);
+
+    uint256 last_header;
+
+    // Filter should not be found in the index before it is started.
+    {
+        LOCK(cs_main);
+
+        BlockFilter filter;
+        uint256 filter_header;
+        std::vector<BlockFilter> filters;
+        std::vector<uint256> filter_hashes;
+
+        for (const CBlockIndex* block_index = chainActive.Genesis();
+             block_index != nullptr;
+             block_index = chainActive.Next(block_index)) {
+            BOOST_CHECK(!filter_index.LookupFilter(block_index, filter));
+            BOOST_CHECK(!filter_index.LookupFilterHeader(block_index, filter_header));
+            BOOST_CHECK(!filter_index.LookupFilterRange(block_index->nHeight, block_index, filters));
+            BOOST_CHECK(!filter_index.LookupFilterHashRange(block_index->nHeight, block_index,
+                                                            filter_hashes));
+        }
+    }
+
+    // BlockUntilSyncedToCurrentChain should return false before index is started.
+    BOOST_CHECK(!filter_index.BlockUntilSyncedToCurrentChain());
+
+    filter_index.Start();
+
+    // Allow filter index to catch up with the block index.
+    constexpr int64_t timeout_ms = 10 * 1000;
+    int64_t time_start = GetTimeMillis();
+    while (!filter_index.BlockUntilSyncedToCurrentChain()) {
+        BOOST_REQUIRE(time_start + timeout_ms > GetTimeMillis());
+        MilliSleep(100);
```

Comment:
14121: Needs rebase and
```suggestion
        UninterruptibleSleep(std::chrono::milliseconds{100});
```